File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        vector<int>ans(n, -1);
        map<int,int>mpp;
        stack<int>st;
        for(int i = m-1; i >= 0; i--){
        while(!st.empty() && st.top() <= nums2[i])st.pop();
        if(st.empty())mpp[nums2[i]] = -1;
        else mpp[nums2[i]] = st.top();
        st.push(nums2[i]);
        }
       for(int i = 0; i < n; i++){
          ans[i] = mpp[nums1[i]];
       }
       return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: list[int], nums2: list[int]) -> list[int]:
        ans = []
        mpp = {}
        st = []
        n = len(nums2)
        for i in range(n-1, -1, -1):
            while(st and st[-1] < nums2[i]):
                st.pop()
            if(st):
                mpp[nums2[i]] = st[-1]
            else:
                mpp[nums2[i]] = -1
            st.append(nums2[i])
        
        m = len(nums1)    
        for i in range(0, m):
            ans.append(mpp[nums1[i]])
        return ans    